lib/utils/input-source-map-tracker.js

```diff
@@ -63,15 +66,26 @@ function fromSource(self, data, whenDone, context) {
       context.files.pop();
     } else if (nextAt == mapMatch.index) {
       var isRemote = /^https?:\/\//.test(sourceMapFile) || /^\/\//.test(sourceMapFile);
+      var isDataUri = DATA_URI.test(sourceMapFile);
       if (isRemote) {
         return fetchMapFile(self, sourceMapFile, context, proceedToNext);
       } else {
         var sourceFile = context.files[context.files.length - 1];
-        var sourceDir = sourceFile ? path.dirname(sourceFile) : self.options.relativeTo;
-        var sourceMapPath = path.resolve(self.options.root, path.join(sourceDir || '', sourceMapFile));
-
-        var sourceMapData = fs.readFileSync(sourceMapPath, 'utf-8');
-        self.trackLoaded(sourceFile || undefined, sourceMapPath, sourceMapData);
+        var sourceMapPath, sourceMapData;
+        if(isDataUri) {
+          // source map's path is the same as the source file it comes from
+          sourceMapPath = path.resolve(self.options.root, sourceFile);
+          var buffer = dataUriToBuffer(sourceMapFile);
```

Comment:
You need the data URI parser to strip off the data URI's mimetype and grab
its charset declaration.  Data URIs, surprisingly, default to ASCII
encoding, so it's important to respect the data URI's charset when parsing
it.  See also: https://github.com/stylus/stylus/issues/1935

On Sun, Jun 14, 2015 at 11:15 AM, Jakub Pawlowicz <notifications@github.com>
wrote:

> In lib/utils/input-source-map-tracker.js
> <https://github.com/jakubpawlowicz/clean-css/pull/600#discussion_r32380660>
> :
>
> >        if (isRemote) {
> >          return fetchMapFile(self, sourceMapFile, context, proceedToNext);
> >        } else {
> >          var sourceFile = context.files[context.files.length - 1];
> > -        var sourceDir = sourceFile ? path.dirname(sourceFile) : self.options.relativeTo;
> > -        var sourceMapPath = path.resolve(self.options.root, path.join(sourceDir || '', sourceMapFile));
> > -
> > -        var sourceMapData = fs.readFileSync(sourceMapPath, 'utf-8');
> > -        self.trackLoaded(sourceFile || undefined, sourceMapPath, sourceMapData);
> > +        var sourceMapPath, sourceMapData;
> > +        if(isDataUri) {
> > +          // source map's path is the same as the source file it comes from
> > +          sourceMapPath = path.resolve(self.options.root, sourceFile);
> > +          var buffer = dataUriToBuffer(sourceMapFile);
>
> Can't we just use new Buffer(sourceMapFile, 'base64').toString()?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jakubpawlowicz/clean-css/pull/600/files#r32380660>.
>